meteor/sofie-server/data-access/repositories/test/mongo-segment-repository.spec.ts

```diff
@@ -0,0 +1,206 @@
+			const mongoConverter: MongoEntityConverter = mock(MongoEntityConverter)
+			const nonExistingId: string = 'nonExistingId'
+			const segment: Segment = createSegment({})
+			await testDatabase.populateDatabaseWithSegments([segment])
+			const db: Db = testDatabase.getDatabase()
+
+			when(mongoConverter.convertSegments(anything())).thenReturn([])
+			const testee: SegmentRepository = await createCommonTestee({
+				mongoConverter: mongoConverter,
+			})
+
+			expect.assertions(2)
+			try {
+				await testee.deleteRundownSegments(nonExistingId)
+			} catch (error) {
+				// eslint-disable-next-line jest/no-conditional-expect
+				expect(error).toBeInstanceOf(DeletionFailedException)
+				// eslint-disable-next-line jest/no-conditional-expect
+				expect(await db.collection(COLLECTION_NAME).countDocuments()).toBe(1)
+			}
+		})
+
+		it('throws exception, when nonexistent rundownId is given', async () => {
+			const expectedErrorMessageFragment: string = 'Expected to delete one or more segments'
+			const mongoConverter: MongoEntityConverter = mock(MongoEntityConverter)
+			const nonExistingId: string = 'nonExistingId'
+			const segment: Segment = createSegment({})
+			await testDatabase.populateDatabaseWithSegments([segment])
+
+			when(mongoConverter.convertSegments(anything())).thenReturn([])
+			const testee: SegmentRepository = await createCommonTestee({
+				mongoConverter: mongoConverter,
+			})
+
+			expect.assertions(2)
+			try {
+				await testee.deleteRundownSegments(nonExistingId)
+			} catch (error) {
+				// eslint-disable-next-line jest/no-conditional-expect
+				expect(error).toBeInstanceOf(DeletionFailedException)
+				// eslint-disable-next-line jest/no-conditional-expect
+				expect((error as DeletionFailedException).message).toContain(expectedErrorMessageFragment)
+			}
+		})
+
+		// eslint-disable-next-line jest/expect-expect
+		it('deletes parts before segments', async () => {
+			const mongoDb: MongoDatabase = mock(MongoDatabase)
+			const partRepository: PartRepository = mock<PartRepository>()
+			const mongoConverter = mock(MongoEntityConverter)
+			const rundownId: string = 'someRundownId'
+			const segment: Segment = createSegment({ rundownId: rundownId })
+			await testDatabase.populateDatabaseWithSegments([segment])
+			const db: Db = testDatabase.getDatabase()
+			const collection = db.collection(COLLECTION_NAME)
+			const spied = spy(collection)
+
+			when(mongoConverter.convertSegments(anything())).thenReturn([segment])
+			when(partRepository.getParts(anything())).thenResolve([])
+			when(mongoDb.getCollection(anything())).thenReturn(collection)
+			const testee: SegmentRepository = await createTestee({
+				mongoConverter: mongoConverter,
+				mongoDb: mongoDb,
+				partRepository: partRepository,
+			})
+
+			await testee.deleteRundownSegments(rundownId)
+
+			verify(partRepository.deleteSegmentParts(anything())).calledBefore(spied.deleteMany(anything()))
+		})
```

Comment:
Why is the order significant? Isn't it just an implementation detail?

Reply:
There are good arguments for either side here.

Since we are using a document driven database then yes, the order isn't technically that important.
But our entities _do_ have a relation to each other and when working with relations it makes a lot of things easier if we delete them in the right order (children before parents etc). 

If you have a business rules that says that you can never delete a Segment while it still has Parts then all of a sudden you would never get "orphaned" Parts and you wouldn't have to worry about that.

A whole different discussion (that I think we should have at some point in the future): Is a document driven database the right option for Sofie?

Reply:
When working with entities like we have, and not using an ORM, it has always felt natural to me to delete children before parent, so that was the way i had implemented it before Lindved mention that i could/should create tests to ensure this order.

That is alteast my original reason for implementing as i did. That it then turns out there is other even good reason for doing it in that order is just great :)